File: two_sum/sum_of_pair_equal_to_target/python/two_sum.py
"""
Sum of pair equal to target.

Given a sorted array (asc) and a target, find if there exists any pair of elements (arr[i], arr[j])
such that their sum is equal to the target.

Input: arr = [10, 20, 35, 50]; target = 70
Output: true ([20, 50] add up to 70)

"""


# two_sum_1 uses the `two pointers` technique.
def two_sum_1(arr: list[int], target: int) -> bool:
    left, right = 0, len(arr) - 1

    while left < right:
        sum = arr[left] + arr[right]

        if sum == target:
            return True
        elif sum < target:
            left += 1
        else:
            right -= 1

    return False


# two_sum_2 uses the dictionary ((hash) map) technique.
def two_sum_2(arr: list[int], target: int) -> bool:
    complements: dict[int, bool] = {}

    for n in arr:
        if n in complements.keys():
            return True

        complement = target - n
        complements[complement] = True

    return False

File: two_sum/sum_of_pair_equal_to_target/lua/two_sum.lua
-- Sum of pair equal to target.
--
-- Given a sorted array (asc) and a target, find if there exists any pair of elements (arr[i],
-- arr[j]) such that their sum is equal to the target.
--
-- Input: arr = [10, 20, 35, 50]; target = 70
-- Output: true ([20, 50] add up to 70)

local function two_sum_1(arr, target)
	local left, right = 1, #arr

	while left < right do
		local sum = arr[left] + arr[right]

		if sum == target then
			return true
		end

		if sum < target then
			left = left + 1
		else
			right = right - 1
		end
	end

	return false
end

local function two_sum_2(arr, target)
	local complements = {}

	for _, n in pairs(arr) do
		if complements[n] ~= nil then
			return true
		end

		complements[target - n] = true
	end

	return false
end

local function test(name, fun)
	local ok, err = pcall(fun)

	if ok then
		print("PASS: " .. name)
	else
		print("FAIL: " .. name)
		print("      " .. err)
		os.exit(1)
	end
end

test("two_sum_1", function()
	local test_cases = {
		{
			arr = {10, 20, 35, 50},
			target = 70,
			expected = true,
		},
		{
			arr = {10, 20, 30},
			target = 70,
			expected = false,
		},
		{
			arr = {-8, 1, 4, 6, 10, 45},
			target = 16,
			expected = true,
		},
	}

	for _, tc in pairs(test_cases) do
		assert(two_sum_1(tc.arr, tc.target) == tc.expected, "Oops!")
	end
end)

test("two_sum_2", function()
	local test_cases = {
		{
			arr = {10, 20, 35, 50},
			target = 70,
			expected = true,
		},
		{
			arr = {10, 20, 30},
			target = 70,
			expected = false,
		},
		{
			arr = {-8, 1, 4, 6, 10, 45},
			target = 16,
			expected = true,
		},
	}

	for _, tc in pairs(test_cases) do
		assert(two_sum_2(tc.arr, tc.target) == tc.expected, "Oops!")
	end
end)
